### ISSUE TITLE
`Dry::Schema::Messages::Abstract#call` may return the wrong result in the event of a hash collision.

### ISSUE THREAD

**Issue by aardvark179** [dry-rb/dry-schema#132]
**Describe the bug**

`Dry::Schema::Messages::Abstract#call`internally caches results based purely on a hash of its arguments. If there is a hash collision then a result may be returned which was calculated for a different set of arguments.

**To Reproduce**

The easiest way to trigger the problem would be to redefine `Hash#hash` to return a more restricted set of values. For example the implementation of `Hash#hash` in Rubinius is roughly
```
def hash
  val = size
  each_pair do |k,v|
    val ^= k.hash
    val ^= v.hash
  end
  val
end
```

**Expected behavior**

A cached value for a different set of arguments should never be returned. If the concern is that the cache might prevent the arguments from being garbage collected then it would be better to cache against `WeakRef.new(args)`.

I will put a change into TruffleRuby to reduce the chance of collisions in our implementation, but that is not a complete fix as collisions could still occur.

**Your environment**

- Affects my production application: **NO**
- Ruby version: TruffleRuby RC17
- OS: Linux

**Comment by solnic** [comment]
> @aardvark179 If there is a hash collision

How can this happen?

**Comment by aardvark179** [comment]
Hash codes are not guaranteed to be unique, the only guarantee you get is that two objects with different hash codes will definitely not be equal.

Hash codes for arrays and hashes are constructed by combining some information about the hash with the hash codes for its keys and values. Since hashes do not require matching entry orders to be equal for the hashes to be considered equal these hash codes must be combined using operations that commute. Unfortunately the algorithm used in TruffleRuby and Rubinius meant that two identical values in a hash canceled out there contributions to the hash code.

I’ve created an internal PR for TruffleRuby to change our strategy for calculating `Hash#hash`, but there still might be collisions, though hopefully far less often.

**Comment by solnic** [comment]
@aardvark179 OK thanks for explaining it to me, this doesn't sound overly optimistic ;) Have you actually encountered a use case where dry-schema would end up returning wrong cached message? We use the same caching strategy (so `args.hash`) in many other places in different rom/dry libs, so if it's not stable and there are better ways I'd love to know them.

**Comment by aardvark179** [comment]
I saw test failures on TruffleRuby until I strengthened our `Hash#hash` implementation. Hash collisions are a real security concern if user controlled data is involved as it can be possible to engineer collisions.

I’m not sure I can suggest a good new strategy as I haven’t looked at all the things you are attempting to use as keys for your caches but you might want to just use them as keys and let concurrent map deal with it. If your key objects are huge, or mutable then you will have 
problems, but they are problems you already had.

**Comment by solnic** [comment]
@aardvark179 this made me realize we have a bug because it was *never* the intention to use mutable, user-generated data to calculate hash keys. The number of possible options is finite, that's why it is safe to cache message templates. Initially, I thought it was only an issue in one specific case, which I fixed in 1.0.1 but then it turned out the same issue exists for all other cases. I'm dealing with this in #133 and it will be addressed in 1.0.2.

**Comment by solnic** [comment]
@aardvark179 I fixed this in `1.0.2`, thanks again for spotting this! I also patched the old version of dry-validation as it had the same issue.

We still have [some failures](https://travis-ci.org/dry-rb/dry-schema/jobs/531449740) though - luckily they crash in the same way, so hopefully a single fix will make the build green on Truffle 😄
